### PR TITLE
refactor: refactor agg unary function

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Refactor the unary agg function:
1. Use one UnaryFunction to create all unary agg functions (this PR: `skewness`, `sum`, `avg`, `min`, `max`, `any`)
2. With `FunctionData` we can refactor `quantile` like functions to this implementation, in next PR.
3. lack of `count` refactor, `count` can be 0 or 1 args and it's calculated independently that diff with other unary functions.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13652)

### CODE REVIEW

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+use common_expression::StateAddr;
+use ethnum::i256;
+
+use crate::aggregates::aggregate_function_factory::AggregateFunctionDescription;
+use crate::aggregates::assert_unary_arguments;
+use crate::aggregates::QuantileStateV2;
+use crate::aggregates::SkewnessStateV2;
+use crate::with_simple_no_number_mapped_type;
+
+pub trait UnaryState<T>: Send + Sync + Clone {
```
Comment (ariesdevil) [inline src/query/functions/src/aggregates/aggregate_unary.rs:43]: `T` here may be no need

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    where Self: Sized;
+}
+
+pub struct AggregateUnaryFunction<S, T>
+where S: UnaryState<T>
+{
+    display_name: String,
+    state: S,
+    _params: Vec<Scalar>,
+    _arguments: Vec<DataType>,
```
Comment (ariesdevil) [inline src/query/functions/src/aggregates/aggregate_unary.rs:74]: no need vector

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    display_name: &str,
+    params: Vec<Scalar>,
+    arguments: Vec<DataType>,
+) -> Result<AggregateFunctionRef> {
+    assert_unary_arguments(display_name, arguments.len())?;
+
+    let data_type = arguments[0].clone();
+
+    match display_name {
+        "skewness_v2" => {
```
Comment (ariesdevil) [inline src/query/functions/src/aggregates/aggregate_unary.rs:207]: for each agg func, should be extract a new function make code clean

File: src/query/functions/src/aggregates/aggregator.rs [hunk 4]
```diff
@@ -134,6 +135,8 @@ impl Aggregators {
             "intersect_count",
             aggregate_bitmap_intersect_count_function_desc(),
         );
+        factory.register("skewness_v2", aggregate_unary_function_desc());
+        factory.register("quantile_disc_v2", aggregate_unary_function_desc());
```
Comment (ariesdevil) [inline src/query/functions/src/aggregates/aggregator.rs:139]: just different names, should be collect names to a vector then batch create

File: src/query/functions/src/aggregates/aggregate_unary.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/functions/src/aggregates/aggregate_unary.rs:43]: Yes.

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+    fn deserialize(&self, reader: &mut &[u8]) -> Result<Self>
+    where Self: Sized;
+}
+
+pub struct AggregateUnaryFunction<S, T>
+where S: UnaryState<T>
+{
+    display_name: String,
+    state: S,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:72]: _state: PhantomData<S>

Comment (ariesdevil) [reply src/query/functions/src/aggregates/aggregate_unary.rs:72]: the `state` is useful now for `accumulate_keys`

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    fn accumulate_keys(
+        &self,
+        places: &[StateAddr],
+        offset: usize,
+        columns: &[Column],
+        _input_rows: usize,
+    ) -> Result<()> {
+        self.state
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:159]: this should be statical dispatched:

```
 for (row, place) in places.iter().enumerate() {
            self.accumulate_row(place.next(offset), columns, row)?;
}
Ok(())
```

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        };
+
+        Ok(Arc::new(func))
+    }
+}
+
+impl<S, T> AggregateFunction for AggregateUnaryFunction<S, T>
+where
+    S: UnaryState<T>,
+    T: Send + Sync,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:118]: T should be `ArgType`

File: src/query/functions/src/aggregates/aggregate_unary.rs (hunk 5 above)
Comment (sundy-li) [reply src/query/functions/src/aggregates/aggregate_unary.rs:63]: agg function is stateless, it won't keep states as field.

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    fn name(&self) -> &str {
+        &self.display_name
+    }
+
+    fn return_type(&self) -> Result<DataType> {
+        Ok(self.return_type.clone())
+    }
+
+    fn init_state(&self, place: StateAddr) {
+        place.write_state(self.state.clone())
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:119]: use `State::default` instead, remove self.state.

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+use crate::aggregates::aggregate_function_factory::AggregateFunctionDescription;
+use crate::aggregates::aggregate_quantile_disc_v2::create_quantile_disc;
+use crate::aggregates::aggregate_skewness_v2::create_skewness;
+use crate::aggregates::assert_unary_arguments;
+
+pub trait UnaryState<T>: Send + Sync + Clone
+where T: ValueType
+{
+    fn add(&mut self, other: Option<T::ScalarRef<'_>>);
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:42]: `fn add(&mut self, other: T::ScalarRef<'_>);`

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[Column],
+        validity: Option<&Bitmap>,
+        _input_rows: usize,
+    ) -> Result<()> {
+        let state: &mut S = place.get::<S>();
+        let column = T::try_downcast_column(&columns[0]).unwrap();
+        state.add_batch(&column, validity)
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:135]: `add_batch`  does not to be a method of state, just use  `state.add(value);` in this function.

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.display_name)
+    }
+}
+
+impl<S, T, R> AggregateUnaryFunction<S, T, R>
+where
+    S: UnaryState<T> + 'static,
+    T: Send + Sync + ValueType + 'static,
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:83]: ValueType is already  with`'static`

File: src/query/functions/src/aggregates/aggregate_skewness_v2.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            return Ok(());
+        }
+        self.n += rhs.n;
+        self.sum += rhs.sum;
+        self.sum_sqr += rhs.sum_sqr;
+        self.sum_cub += rhs.sum_cub;
+        Ok(())
+    }
+
+    fn merge_result(&mut self) -> Result<Option<Scalar>> {
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_skewness_v2.rs:88]: fn merge_result<R:  ValueType>(&mut self, StateAddr,  _place: StateAddr, _builder: &mut R::ColumnBuilder) -> Result<()> {

File: src/query/functions/src/aggregates/aggregate_unary.rs [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    display_name: &str,
+    params: Vec<Scalar>,
+    arguments: Vec<DataType>,
+) -> Result<AggregateFunctionRef> {
+    assert_unary_arguments(display_name, arguments.len())?;
+
+    let data_type = arguments[0].clone();
+
+    match display_name {
+        "skewness_v2" => create_skewness(display_name, params, &arguments, &data_type),
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_unary.rs:239]: null adaptor is missing.

Comment (sundy-li) [reply src/query/functions/src/aggregates/aggregate_unary.rs:239]: should update `aggregate_unary_function_desc`

File: src/query/functions/tests/it/aggregates/testdata/agg.txt [hunk 14]
```diff
@@ -740,7 +740,7 @@ evaluation (internal):
 | Column | Data                                                                    |
 +--------+-------------------------------------------------------------------------+
 | x_null | NullableColumn { column: UInt64([1, 2, 3, 4]), validity: [0b____0011] } |
-| Output | NullableColumn { column: Float64([0]), validity: [0b_______0] }         |
+| Output | NullableColumn { column: Float64([0]), validity: [0b_______1] }         |
```
Comment (sundy-li) [inline src/query/functions/tests/it/aggregates/testdata/agg.txt:743]: is this change ok?

Comment (ariesdevil) [reply src/query/functions/tests/it/aggregates/testdata/agg.txt:743]: Yup, the original implementation handles the nullable manually, so this diff happens.

File: src/query/functions/src/aggregates/aggregate_avg.rs [hunk 15]
```diff
@@ -12,185 +12,245 @@
...
+{
+    pub value: T::Scalar,
+    pub count: u64,
+}
 
-        serialize_state(writer, state)
+impl<const OVERFLOW: bool, T> Default for DecimalAvgState<OVERFLOW, T>
+where
+    T: ValueType,
+    T::Scalar: Decimal
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_avg.rs:139]: Decimal trait covers other trait, need deduplication

File: src/query/functions/src/aggregates/aggregate_skewness.rs [hunk 16]
```diff
@@ -1,272 +1,161 @@
...
+        let temp1 = (n * (n - 1.0)).sqrt() / (n - 2.0);
+        let value = temp1
+            * temp
+            * (self.sum_cub - 3.0 * self.sum_sqr * self.sum * temp
+                + 2.0 * self.sum.powi(3) * temp * temp)
+            / div;
+        if value.is_infinite() || value.is_nan() {
+            return Err(ErrorCode::SemanticError("Skew is out of range!"));
+        } else {
+            R::push_item(
```
Comment (sundy-li) [inline src/query/functions/src/aggregates/aggregate_skewness.rs:115]: R is f64 already

**Review by sundy-li — APPROVED** [review]

**Pull request merged**